Scientist4JCore/src/main/java/com/github/rawls238/scientist4j/Experiment.java

```diff
@@ -59,77 +62,106 @@ public String getName() {
     }
 
     public T run(Supplier<T> control, Supplier<T> candidate) throws Exception {
-      Observation<T> controlObservation;
-      Optional<Observation> candidateObservation = Optional.empty();
-      if (Math.random() < 0.5) {
-        controlObservation = executeResult("control", controlTimer, control, true);
-        if (runIf() && enabled()) {
-          candidateObservation = Optional.of(executeResult("candidate", candidateTimer, candidate, false));
+        Observation<T> controlObservation;
+        Optional<Observation> candidateObservation = Optional.empty();
+        if (isAsync()) {
+            return runAsync(control, candidate);
+        } else {
+            if (Math.random() < 0.5) {
+                controlObservation = executeResult("control", controlTimer, control, true);
+                if (runIf() && enabled()) {
+                    candidateObservation = Optional.of(executeResult("candidate", candidateTimer, candidate, false));
+                }
+            } else {
+                if (runIf() && enabled()) {
+                    candidateObservation = Optional.of(executeResult("candidate", candidateTimer, candidate, false));
+                }
+                controlObservation = executeResult("control", controlTimer, control, true);
+            }
+            Result<T> result = new Result(this, controlObservation, candidateObservation, context);
+            publish(result);
+            return controlObservation.getValue();
         }
-      } else {
+    }
+
+    public T runAsync(Supplier<T> control, Supplier<T> candidate) throws Exception {
+        Future<Observation> observationFutureCandidate = null;
+        Future<Observation<T>> observationFutureControl = executor.submit(() -> executeResult("control", controlTimer, control, true));
```

Comment:
this matters less because they're being run in parallel but we should still randomize between which one gets dispatched first